Add support for PUT upload

eXenon opened the pull request:
- Add support for the `method` field returned by CAP instances
- Add support for the PUT upload method
- Add `--no-check-certificate` option to disable certificate validation

eXenon committed:
Add option to give a raw PUT url.

Also add option to remove certificate check

eXenon committed:
Don't use multipart form with PUT

eXenon committed:
Add compatibility with updated API

eXenon committed:
Fix CI

eXenon committed:
Formating

eXenon committed:
Fix typing

sbq-clements reviewed (approved):
Seems OK to me 👍.

bertrand-sb commented:
@eXenon Could you update this now that the server advertises its preferred upload method?

eXenon committed:
Remove deprecated put_url option

eXenon committed:
Fix formatting

eXenon commented:
> @eXenon Could you update this now that the server advertises its preferred upload method?

Added! I had done most of the work in anticipation, so it was only a matter of removing the "bridge" options I had added.

eXenon committed:
Fix small mistakes

bertrand-sb commented on cs_upload.py:
```diff
@@ -362,7 +458,12 @@ def main():
-        help="Trace or scan file to upload.",
+        help="Print out more detailed logging",
+    )
+    parser.add_argument(
+        "--insecure",
+        action="store_true",
+        help="Disable the certificate validation.",
```
**suggestion**: Use `--no-check-certificate` since this is what we have in the OCaml cs-api tool.

Also, I would suggest using this name internally instead of `check_ca_cert` which implies we only check the CA cert, which is incorrect: curl also checks the whole chain and the validity of the server certificate (hostname + expiration + other properties).

eXenon replied:
Changed!

bertrand-sb commented on cs_upload.py:
```diff
@@ -286,51 +334,99 @@ class S3Client:
+        object_storage_url: str,
+        ca_cert: Optional[str],
+        check_ca_cert: bool = True,
+        upload_method: str = "POST",
```
**suggestion**: Remove the default value for `upload_method` since we always define it in our script (and it doesn't seem to me like `POST` should be the preferred one anyway).

bertrand-sb commented on cs_upload.py:
```diff
@@ -286,51 +334,99 @@ class S3Client:
         else:
             print(
-                "Trace file extension must be either .cst.gz or .cst",
+                "Trace file extension must be either .pcap, .cst.gz or .cst",
```
**suggestion**: Add `.pcap.gz`, assuming we support that format (I think I saw one somewhere).

eXenon replied:
No, we don't support that format in the end. I thought we could, but implementation in OCaml to detect that format turned out to be a nightmare, so I decided against it. For now at least.

bertrand-sb commented on cs_upload.py:
```diff
@@ -286,51 +334,99 @@ class S3Client:
             )
             exit(1)
 
+    def upload_to_s3(self, form_data: Optional[str], trace_file: str) -> None:
```
**suggestion**: Remove the `Optional` for `form_data`. I don't understand how it can be `None`.

bertrand-sb commented on cs_upload.py:
```diff
@@ -120,26 +150,44 @@ def query(
         )
 
-    def generate_trace_upload_post(self) -> Tuple[str, str]:
+    def generate_trace_upload(self) -> Tuple[str, str, str]:
```
**suggestion**: Define a dataclass:

```python

class UploadMethod(enum.Enum):
    PUT = enum.auto()
    POST = enum.auto()

@dataclass(frozen=True)
class UploadInfo:
    object_storage_url: str
    form_data: str
    method: UploadMethod
```

I also suggest parsing the HTTP method here:

- to detect a server issue early: If there's such an error, it's better to abort before any other action is taken by our script.
- to avoid having to check for the right string every time we use this parameter (I found 3 occurrences of the "POST").

eXenon replied:
dataclasses will very slightly reduce our compatibility but that is probably fine (python 3.6 was the minimum before and with dataclasses, it moves to 3.7 IIRC).

bertrand-sb commented on cs_upload.py:
```diff
@@ -74,15 +74,41 @@ class GraphQLClient:
+        ]
+        return run(query, capture_output=True)
+
+    def _put(self, data: str) -> CompletedProcess:
```
**suggestion**: Remove this method. Do we really need to use `PUT` for GraphQL queries?

eXenon replied:
Good catch, removed it.

bertrand-sb reviewed (changes requested):
Thanks for the update. There are things I don't understand so I'm looking forward to the next iteration.

eXenon committed:
Integrate review comments

bertrand-sb reviewed (approved):
Thanks for the update!
